### PR TITLE
eval harness integration

### PR DESCRIPTION
**Pull request by aniketmaurya**
Implements eval harness adapter for Lit-GPT.

```bash
python eval/eval_adapter.py \
        --checkpoint_dir "checkpoints/Llama-2-7b-hf/" \
        --precision "bf16-true" \
        --eval_tasks "[truthfulqa_mc]" \
        --batch_size 4
```

_Few benchmarks using this PR (Matches closely with the LLM leaderboard):_
image

### CODE REVIEW

**Comment by rasbt** [comment]
Awesome work, Aniket! Just a spontaneous thought: could people potentially confuse it with "evaluate adapter-finetuned LLM" since we have `generate/generate_adapter.py`. 

Should we rename `python eval/eval_adapter.py` to `python eval/eval_harness.py` @aniketmaurya @lantiga?

(And then may have `python eval/hellaswag.py` etc in the future?)

**Comment by aniketmaurya** [comment]
Thank you Sebastian! I agree with you. I'll rename the module to eval_harness.py to avoid any confusion.

File: tutorials/evaluation.md [hunk 1]
```diff
@@ -0,0 +1,23 @@
+# LLM Evaluation
+
+## Using lm-evaluation-harness
+
+You can evaluate Lit-GPT using [EleutherAI's lm-eval](https://github.com/EleutherAI/lm-evaluation-harness/tree/master) framework large number of different evaluation tasks.
```
Comment (Andrei-Aksionov) [inline tutorials/evaluation.md:5]:
```suggestion
You can evaluate Lit-GPT using [EleutherAI's lm-eval](https://github.com/EleutherAI/lm-evaluation-harness/tree/master) framework with a large number of different evaluation tasks.
```

File: eval/eval_adapter.py [hunk 2]
```diff
@@ -0,0 +1,362 @@
...
+        self.vocab_size = self.tokenizer.vocab_size
+
+        # multithreading and batching
+        self.batch_size_per_gpu = batch_size
+
+        self.temperature = temperature
+
+    @classmethod
+    def create_from_arg_string(cls, arg_string, additional_config=None):
+        kwargs = {el.split("=")[0]: el.split("=")[1] for el in arg_string.split(",")}
```
Comment (Andrei-Aksionov) [inline eval/eval_adapter.py:179]: Interestingly enough this one should work (though I haven't run your code 🙃 ):
```python
kwargs = dict(el.split("=") for el in arg_string.split(","))
```

Comment (Andrei-Aksionov) [reply eval/eval_adapter.py:179]: Plus it's a bit more performant as there is no need to split each `el` twice. 
Python go brrrr!

Screenshot 2023-08-06 at 6 53 32 PM

Just a Sunday joke 😄

File: eval/eval_adapter.py [hunk 3]
```diff
@@ -0,0 +1,362 @@
...
+    )
+    adapter.fabric.print("Running evaluation harness...")
+    results = adapter.run_eval(
+        eval_tasks=eval_tasks,
+        num_fewshot=num_fewshot,
+        bootstrap_iters=bootstrap_iters,
+        use_cache=False,
+    )
+    data = json.dumps(results)
+    filename = str(time.time()) + "-eval.txt"
```
Comment (Andrei-Aksionov) [inline eval/eval_adapter.py:345]: Is it a convention to use that date/time format? Maybe some OS might sort weirdly if the timing is not in this format (how it's called, ticks perhaps)? Or maybe then the date/time is parsed from the name of the file?
I know it's a small, really-really small issue, but I prefer to see more obvious date/time format in the filename:

Screenshot 2023-08-06 at 7 06 15 PM

Comment (aniketmaurya) [reply eval/eval_adapter.py:345]: `time.time()` value is same everywhere in the world when called at the same moment. my intent here was only to have a unique filename (sortable with timestamp).

File: eval/eval_adapter.py (hunk 2 above)
Comment (aniketmaurya) [reply eval/eval_adapter.py:179]: nice find! I will do some cleanup in the PR tomorrow morning 😄

**Review by lantiga — APPROVED** [review]
LGTM!

**Pull request merged**